Check for WebView2 Runtime support during launch.

**Describe the bug**
When launching the application without WebView2 runtime support, the app closes immediately upon opening.

**To Reproduce**
Launch application on an unsupported configuration; lacking WebView2 runtimes.

**Expected behavior**
Per Microsoft - "When distributing your WebView2 app, ensure the backing web platform, the WebView2 Runtime, is present before the app starts."
https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution

**Screenshots**
None required.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Any
 - Version: 8.2.2

**Smartphone (please complete the following information):**
N/A

**Additional context**
Log:
02:45:11:661 => [ERROR] [WolvenKit.Functionality.Helpers.StaticReferences] [8] [WebView2RuntimeNotFoundException] Microsoft.Web.WebView2.Core.WebView2RuntimeNotFoundException: Couldn't find a compatible Webview2 Runtime installation to host WebViews.

--- a/WolvenKit.App/Functionality/Helpers/Debug_Helpers.cs
+++ b/WolvenKit.App/Functionality/Helpers/Debug_Helpers.cs
@@ -7,11 +7,6 @@ using System.Threading.Tasks;
 
 namespace WolvenKit.Functionality.Helpers
 {
-    public static partial class VMHelpers
-    {
-        public static bool InDebug { get; set; } = false;
-    }
-
     public static class WolvenDBG
     {
         public static bool EnableTheming { get; set; } = true;

--- a/WolvenKit.App/Models/Commonfunctions.cs
+++ b/WolvenKit.App/Models/Commonfunctions.cs
@@ -8,12 +8,44 @@ using System.Xml;
 using System.Xml.Linq;
 using ICSharpCode.SharpZipLib.Core;
 using ICSharpCode.SharpZipLib.Zip;
+using Microsoft.Win32;
 
 namespace WolvenKit.Models
 {
     public static class Commonfunctions
     {
-        #region Methods
+
+
+        /// <summary>
+        /// Check if a registry value exists
+        /// </summary>
+        /// <param name="hive"></param>
+        /// <param name="registryRoot"></param>
+        /// <param name="valueName"></param>
+        /// <returns></returns>
+        public static bool RegistryValueExists(RegistryHive hive, string registryRoot, string valueName)
+        {
+            RegistryKey root;
+            switch (hive)
+            {
+                case RegistryHive.LocalMachine:
+                    root = Registry.LocalMachine.OpenSubKey(registryRoot, false);
+                    break;
+                case RegistryHive.CurrentUser:
+                    root = Registry.CurrentUser.OpenSubKey(registryRoot, false);
+                    break;
+                case RegistryHive.ClassesRoot:
+                case RegistryHive.Users:
+                case RegistryHive.PerformanceData:
+                case RegistryHive.CurrentConfig:
+                default:
+                    throw new System.InvalidOperationException("parameter registryRoot must be either \"HKLM\" or \"HKCU\"");
+            }
+
+            var r = root.GetValue(valueName) != null;
+            return r;
+        }
+
 
         /// <summary>
         /// Compresses a file into a zipstream.
@@ -289,6 +321,5 @@ namespace WolvenKit.Models
             return ms.ToArray();
         }
 
-        #endregion Methods
     }
 }

--- a/WolvenKit.App/ViewModels/Shared/RecentlyUsedItemsViewModel.cs
+++ b/WolvenKit.App/ViewModels/Shared/RecentlyUsedItemsViewModel.cs
@@ -126,11 +126,10 @@ namespace WolvenKit.ViewModels.Shared
 
         private void ConvertRecentProjects() // Converts Recent projects for the homepage.
         {
-            //DispatcherHelper.RunOnMainThread(() =>
-            //           {
-            //               FancyProjects.Clear();
-            //           });
-            FancyProjects.Clear();
+            DispatcherHelper.RunOnMainThread(() =>
+                       {
+                           FancyProjects.Clear();
+                       });
 
             foreach (var item in _recentlyUsedItems)
             {

--- a/WolvenKit.App/ViewModels/Shell/WorkSpaceViewModel.cs
+++ b/WolvenKit.App/ViewModels/Shell/WorkSpaceViewModel.cs
@@ -162,6 +162,7 @@ namespace WolvenKit.ViewModels.Shell
             });
 
             await _updateService.CheckForUpdatesAsync();
+
         }
 
         protected override async Task InitializeAsync()

--- a/WolvenKit.Core/CommonFunctions.cs
+++ b/WolvenKit.Core/CommonFunctions.cs
@@ -7,6 +7,7 @@ using System.Runtime.InteropServices;
 using System.Security.Cryptography;
 using System.Text;
 using System.Threading.Tasks;
+using Microsoft.Win32;
 
 namespace WolvenKit.Core
 {
@@ -65,5 +66,7 @@ namespace WolvenKit.Core
 
         // Display a byte array in a readable format.
         public static string PrettyByteArray(IEnumerable<byte> array) => array.Aggregate("", (current, t) => current + $"{t:X2}");
+
+        
     }
 }

--- a/WolvenKit.Core/WolvenKit.Core.csproj
+++ b/WolvenKit.Core/WolvenKit.Core.csproj
@@ -26,6 +26,7 @@
         <PackageReference Include="DotNetHelper.FastMember.Extension" Version="2.0.0" />
         <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.10.0" />
         <PackageReference Include="CsvHelper" Version="27.1.1" />
+        <PackageReference Include="Microsoft.Win32.Registry" Version="5.0.0" />
         <PackageReference Include="protobuf-net.Core" Version="3.0.101" />
         <PackageReference Include="ReactiveUI" Version="14.1.1" />
         <PackageReference Include="System.Reflection.MetadataLoadContext" Version="5.0.1" />

--- a/WolvenKit.sln
+++ b/WolvenKit.sln
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29509.3
+# Visual Studio Version 17
+VisualStudioVersion = 17.0.31521.260
 MinimumVisualStudioVersion = 10.0.40219.1
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "_Solution_Items", "_Solution_Items", "{6E63DB95-528E-4816-BE98-7F339599C512}"
 	ProjectSection(SolutionItems) = preProject
@@ -36,7 +36,7 @@ EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Installers", "Installers", "{C22F3C33-7D76-48BC-97DA-95F0BA300775}"
 	ProjectSection(SolutionItems) = preProject
 		Installers\installer.iss = Installers\installer.iss
-		Installers\setup.aip = Installers\setup.aip
+		Installers\wkit_setup.aip = Installers\wkit_setup.aip
 	EndProjectSection
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WolvenKit", "WolvenKit\WolvenKit.csproj", "{283EA282-F15F-4978-8FF4-43EFE7A9EF69}"

--- a/WolvenKit/App.xaml.cs
+++ b/WolvenKit/App.xaml.cs
@@ -46,18 +46,16 @@ namespace WolvenKit
         // Constructor #2
         public App()
         {
-            // SPLAT
             Init();
 
+            ConfigureServices();            
+        }
 
-            // Set application licenses.
-            Initializations.InitializeLicenses();
-            //protobuf
-            RuntimeTypeModel.Default[typeof(IGameArchive)].AddSubType(20, typeof(Archive));
-
+        private void ConfigureServices()
+        {
             var serviceLocator = ServiceLocator.Default;
             serviceLocator.RegisterType<INotificationService, NotificationService>();
-            
+
             var config = SettingsManager.Load();
             serviceLocator.RegisterInstance(typeof(ISettingsManager), config);
 
@@ -65,7 +63,7 @@ namespace WolvenKit
 
             //serviceLocator.RegisterType<ILoggerService, CatelLoggerService>();
             serviceLocator.RegisterType<ILoggerService, ReactiveLoggerService>();
-            
+
             serviceLocator.RegisterType<IUpdateService, UpdateService>();
 
             // singletons
@@ -100,8 +98,14 @@ namespace WolvenKit
         }
 
         // Application OnStartup Override.
-        protected override /*async*/ void OnStartup(StartupEventArgs e)
+        protected override void OnStartup(StartupEventArgs e)
         {
+            // check prerequisites
+            // check Webview2
+            string keyName = @"SOFTWARE\Wow6432Node\Microsoft\EdgeUpdate\ClientState\{F3017226-FE2A-4295-8BDF-00C3A9A7E4C5}";
+            string keyvalue = "pv";
+            StaticReferences.IsWebView2Enabled = Models.Commonfunctions.RegistryValueExists(Microsoft.Win32.RegistryHive.LocalMachine, keyName, keyvalue);
+
             // Startup speed boosting (HC)
             ApplicationHelper.StartProfileOptimization();
 
@@ -110,7 +114,6 @@ namespace WolvenKit
 
 #if DEBUG
             LogManager.AddDebugListener(false);
-            VMHelpers.InDebug = true;
 #endif
 
             StaticReferences.Logger.Info("Starting application");
@@ -235,7 +238,7 @@ namespace WolvenKit
             _commandManager.CreateCommand(AppCommands.Application.ViewSelected);
         }
 
-        protected override /*async*/ void OnExit(ExitEventArgs e)
+        protected override void OnExit(ExitEventArgs e)
         {
             base.OnExit(e);
         }
@@ -246,6 +249,12 @@ namespace WolvenKit
         //private IHost _host;
         void Init()
         {
+            // Set application licenses.
+            Initializations.InitializeLicenses();
+            //protobuf
+            RuntimeTypeModel.Default[typeof(IGameArchive)].AddSubType(20, typeof(Archive));
+
+
             //_host = Host
             //    .CreateDefaultBuilder()
             //    .ConfigureServices(services =>
@@ -261,7 +270,7 @@ namespace WolvenKit
             //    //})
             //    .ConfigureServices((hostContext, services) =>
             //    {
-                   
+
 
 
             //        // this passes IScreen resolution through to the previous viewmodel registration.
@@ -270,7 +279,7 @@ namespace WolvenKit
             //        //services.AddSingleton<IScreen, WorkSpaceViewModel>(x => x.GetRequiredService<WorkSpaceViewModel>());
             //        services.AddSingleton<IViewFor<WorkSpaceViewModel>, MainView>();
 
-                   
+
 
 
             //    })

--- a/WolvenKit/Functionality/Helpers/Static_Helpers.cs
+++ b/WolvenKit/Functionality/Helpers/Static_Helpers.cs
@@ -19,6 +19,8 @@ namespace WolvenKit.Functionality.Helpers
 
         public static bool AllowVideoPreview { get; set; } = true;
 
+        public static bool IsWebView2Enabled { get; set; } = false;
+
         /// <summary>
         /// Clean settings below this text.
         /// </summary>

--- a/WolvenKit/Functionality/Initialization/Initializations.cs
+++ b/WolvenKit/Functionality/Initialization/Initializations.cs
@@ -159,6 +159,11 @@ namespace WolvenKit.Functionality.Initialization
         /// </summary>
         public async static void InitializeWebview2()
         {
+            if (!StaticReferences.IsWebView2Enabled)
+            {
+                return;
+            }
+
             var webViewData = ISettingsManager.GetWebViewDataPath();
             Directory.CreateDirectory(webViewData);
             Helpers.Helpers.objCoreWebView2Environment = await CoreWebView2Environment.CreateAsync(null, webViewData, null);

--- a/WolvenKit/Views/HomePage/Pages/GithubPageView.xaml
+++ b/WolvenKit/Views/HomePage/Pages/GithubPageView.xaml
@@ -15,7 +15,7 @@
      >
         <Grid Margin="3,75,3,0">
 
-                <wv2:WebView2 x:Name="cal" Source="https://github.com/WolvenKit/Wolven-kit" />
+                <wv2:WebView2 x:Name="cal"/>
         </Grid>
 
             <StackPanel

--- a/WolvenKit/Views/HomePage/Pages/GithubPageView.xaml.cs
+++ b/WolvenKit/Views/HomePage/Pages/GithubPageView.xaml.cs
@@ -16,6 +16,11 @@ namespace WolvenKit.Views.HomePage.Pages
 
         private async void dome()
         {
+            if (!StaticReferences.IsWebView2Enabled)
+            {
+                return;
+            }
+
             try
             {
                 await cal.EnsureCoreWebView2Async(Helpers.objCoreWebView2Environment);

--- a/WolvenKit/Views/HomePage/Pages/WebsitePageView.xaml
+++ b/WolvenKit/Views/HomePage/Pages/WebsitePageView.xaml
@@ -16,7 +16,7 @@
 
         <Grid Margin="3,75,3,0">
 
-                <wv2:WebView2 x:Name="cal" Source="https://redmodding.org/" />
+                <wv2:WebView2 x:Name="cal" />
         </Grid>
 
             <StackPanel

--- a/WolvenKit/Views/HomePage/Pages/WebsitePageView.xaml.cs
+++ b/WolvenKit/Views/HomePage/Pages/WebsitePageView.xaml.cs
@@ -16,6 +16,11 @@ namespace WolvenKit.Views.HomePage.Pages
 
         private async void dome()
         {
+            if (!StaticReferences.IsWebView2Enabled)
+            {
+                return;
+            }
+
             try
             {
                 await cal.EnsureCoreWebView2Async(Helpers.objCoreWebView2Environment);

--- a/WolvenKit/Views/HomePage/Pages/WelcomePageView.xaml
+++ b/WolvenKit/Views/HomePage/Pages/WelcomePageView.xaml
@@ -462,12 +462,12 @@
                     </StackPanel>
                     <hc:Divider Margin="15,5,15,3" />
 
-                    <wv2:WebView2
+                    <wv2:WebView2 x:Name="cal"
                         Width="414"
                         Height="{Binding ElementName=lokz, Path=ActualHeight}"
                         MaxWidth="414"
                         Margin="0,0,15,0"
-                        Source="https://www.youtube.com/embed/WfEi3QwhTIs" />
+                        />
                 </StackPanel>
             </Border>
         </Grid>

--- a/WolvenKit/Views/HomePage/Pages/WelcomePageView.xaml.cs
+++ b/WolvenKit/Views/HomePage/Pages/WelcomePageView.xaml.cs
@@ -1,3 +1,6 @@
+using System;
+using WolvenKit.Functionality.Helpers;
+
 namespace WolvenKit.Views.HomePage.Pages
 {
     public partial class WelcomePageView
@@ -5,6 +8,26 @@ namespace WolvenKit.Views.HomePage.Pages
         public WelcomePageView()
         {
             InitializeComponent();
+            dome();
+        }
+
+        private async void dome()
+        {
+            if (!StaticReferences.IsWebView2Enabled)
+            {
+                return;
+            }
+
+            try
+            {
+                await cal.EnsureCoreWebView2Async(Helpers.objCoreWebView2Environment);
+
+                cal.SetCurrentValue(Microsoft.Web.WebView2.Wpf.WebView2.SourceProperty, (System.Uri)new Uri("https://www.youtube.com/embed/WfEi3QwhTIs"));
+            }
+            catch (Exception)
+            {
+                // TODO: handle this
+            }
         }
     }
 }

--- a/WolvenKit/Views/HomePage/Pages/WikiPageView.xaml.cs
+++ b/WolvenKit/Views/HomePage/Pages/WikiPageView.xaml.cs
@@ -17,6 +17,11 @@ namespace WolvenKit.Views.HomePage.Pages
         // initialize webviev for Wiki page.
         private async void InitializeWebviev()
         {
+            if (!StaticReferences.IsWebView2Enabled)
+            {
+                return;
+            }
+
             try
             {
                 await cal.EnsureCoreWebView2Async(Helpers.objCoreWebView2Environment);

--- a/WolvenManager.Installer/Services/UpdateService.cs
+++ b/WolvenManager.Installer/Services/UpdateService.cs
@@ -4,6 +4,7 @@ using System.ComponentModel;
 using System.IO;
 using System.Net;
 using System.Net.Http;
+using System.Net.NetworkInformation;
 using System.Reactive.Linq;
 using System.Security.Cryptography;
 using System.Text.Json;
@@ -84,6 +85,23 @@ namespace WolvenManager.Installer.Services
                 return;
             }
 
+            // ping
+            try
+            {
+                var host = GetManifestUri().Host;
+                var reply = new Ping().Send(host, 3000);
+                if (reply.Status != IPStatus.Success)
+                {
+                    return;
+                }
+            }
+            catch (Exception ex)
+            {
+
+                _loggerService.Error(ex.Message);
+                return;
+            }
+
             // check versions
             var http = new HttpClient();
             var manifestJson = await http.GetStringAsync(GetManifestUri());
@@ -109,6 +127,7 @@ namespace WolvenManager.Installer.Services
                     await HandleUpdate(EIncludedFiles.Portable);
                 }
             }
+
 
             async Task HandleUpdate(EIncludedFiles type)
             {

